### DISCUSSION TITLE
Multi-threading when exporting mesh?

### DISCUSSION THREAD

**Discussion by zsfab** [curv3d/curv#111] [Ideas]
I would like to ask whether it is possible to do multi-threading when exporting mesh? I have a computer with many cores and would like to utilize it to the most when running curv, I am just wondering.

**Comment by doug-moen** [comment]
This is not implemented yet.

If anybody wants to contribute this, please talk to me before submitting a pull request.

**Reply by Xiaoyuew** [reply]
I would like to contribute, could you inform me of some more details?

**Reply by doug-moen** [reply]
I think this is a small, simple project with high impact.
You need to modify the function `export_mesh`, in the file `curv/export_mesh.cc`.

When you export a mesh using the recommended `-O jit` option, the algorithm proceeds in 3 steps:
1. Compile the shape's colour and dist functions into C++ code, which is compiled by the C++ compiler, then the resulting machine code is loaded into memory. There is no easy way to speed this up.
2. Create a 3 dimensional array, a voxel grid, by sampling the distance function. Calling the distance function millions of times to construct the voxel grid is where most of the time is spent. This is the part that is easiest to speed up using multiple cores.
3. Convert the voxel grid to a triangle mesh using an external library. I currently use openvdb, but other libraries exist. Whether this step makes use of multiple cores depends on the choice of mesh generation library. I am interested in generalizing this step so that the user can choose between several libraries using a command line option, because all the libraries make different tradeoffs in speed and output quality and what types of models they work best with, but that is a different project. 

So, the goal is to speed up creation of the voxel grid by using multiple cores. There are two variants of this code: one uses the Curv interpreter, which is very slow, and also not thread safe (so it cannot be sped up using multiple cores). The other variant uses the compiled output of the C++ compiler. This is the `-Ojit` option, and in this variant, the distance function *is* thread safe, so you can call it in parallel using multiple cores.

To create a voxel grid, you must construct a 3D array variable, then populate the array. At present, this code uses openvdb to create the grid. I seem to recall that the openvdb API calls I am using are not thread safe, so it is difficult to parallelize, but you could research openvdb and see if there is a recommended way to do this that uses multiple cores. I don't necessarily recommend this approach.

I think the simplest approach is:
1. Create a regular C++ array of `float` that contains `x*y*z` elements.
2. Populate this array by sampling the distance distance function. Use multiple cores to speed this up.
3. Then, copy the data from the C++ array into the openvdb voxel grid. This copying introduces an extra step, but hopefully it only requires a small fraction of second, a slowdown that won't be noticed.

In this simple approach, the loop that populates the voxel grid is 7 lines of code: it is 3 for loops, nested, with a call to the distance function in the innermost loop. I think you can use *openmp* to distribute this across multiple cores. I think you will add a `#pragma omp parallel for` line to run the outermost for loop in parallel. That should just be 1 extra line of code. I've never used openmp, but you can research it.

**Reply by Xiaoyuew** [reply]
Thank you! Will look into it!

**Reply by Xiaoyuew** [reply]
Hi Doug, I think it would be nice to add the feature that the mesh could be exported to binary format.

**Reply by doug-moen** [reply]
Are you referring to binary STL, or to another binary format?
I'm not planning to work on this feature right now, but if you want to make a contribution, I will accept it.
The first step is create an issue describing the new feature. Once the pull request is published, the issue will be marked as resolved.

**Comment by Xiaoyuew** [comment]
Yes binary STL. I will consider working on this :)
